src/sage/data_structures/pairing_heap.pyx

```diff
@@ -0,0 +1,1523 @@
+            sage: P.push(2, 3); P
+            PairingHeap_of_n_integers: capacity 5, size 2
+            sage: P.pop(); P
+            PairingHeap_of_n_integers: capacity 5, size 1
+            sage: P.pop(); P
+            PairingHeap_of_n_integers: capacity 5, size 0
+            sage: P.pop(); P
+            PairingHeap_of_n_integers: capacity 5, size 0
+        """
+        if not self:
+            return
+        cdef size_t item = self.top_item()
+        Py_XDECREF(<PyObject *>self.nodes[item].value)
+        bitset_remove(self.active, item)
+        self.number_of_items -= 1
+        self.root = _pair(self.root.child)
+
+    cpdef void decrease(self, size_t item, object new_value) except *:
+        r"""
+        Decrease the value of specified item.
+
+        This method is more permissive than it should as it can also be used to
+        push an item in the heap.
+
+        INPUT:
+
+        - ``item`` -- non negative integer; the item to consider
+
+        - ``new_value`` -- the new value for ``item``
+
+        EXAMPLES::
+
+            sage: from sage.data_structures.pairing_heap import PairingHeap_of_n_integers
+            sage: P = PairingHeap_of_n_integers(5)
+            sage: 3 in P
+            False
+            sage: P.decrease(3, 33)
+            sage: 3 in P
+            True
+            sage: P.top()
+            (3, 33)
+            sage: P.push(1, 10)
+            sage: P.top()
+            (1, 10)
+            sage: P.decrease(3, 7)
+            sage: P.top()
+            (3, 7)
+
+        TESTS::
+
+            sage: P = PairingHeap_of_n_integers(5)
+            sage: P.push(1, 3)
+            sage: P.decrease(1, 2)
+            sage: P.decrease(1, 2)
+            Traceback (most recent call last):
+            ...
+            ValueError: the new value must be less than the current value
+        """
+        cdef PairingHeapNode * p
+        if bitset_in(self.active, item):
+            p = self.nodes + item
+            if <object>p.value <= new_value:
+                raise ValueError("the new value must be less than the current value")
+            Py_XDECREF(<PyObject *>p.value)
+            Py_INCREF(new_value)
+            p.value = <void *>new_value
+            if p.prev != NULL:
+                _unlink(p)
+                self.root = _merge(self.root, p)
+        else:
+            self.push(item, new_value)
+
+    def __contains__(self, size_t item):
+        r"""
+        Check whether the specified item is in the heap.
+
+        INPUT:
+
+        - ``item`` -- non negative integer; the item to consider
+
+        EXAMPLES::
+
+            sage: from sage.data_structures.pairing_heap import PairingHeap_of_n_integers
+            sage: P = PairingHeap_of_n_integers(5)
+            sage: 3 in P
+            False
+            sage: P.push(3, 33)
+            sage: 3 in P
+            True
+            sage: 100 in P
+            False
+        """
+        return bitset_in(self.active, item)
```

Comment:
Is a check needed before the call to `bitset_in` to be sure that `item` is less than the size of the bitset ? The code of  `bitset_in` does not seem to perform any check.

Reply:
I will add it